File: src/main/java/medium/_518_coin_change_2/Solution.java
package medium._518_coin_change_2;

// DP top-down
public class Solution implements ISolution {
    public int change(int amount, int[] coins)  {
        if (amount == 0) {
            return 1;
        }
        if (coins.length ==0) {
            return 0;
        }

        int[][] memo = new int[amount+1][coins.length];
        for (int i =0; i < memo.length; i++)
            for (int j =0; j < memo[0].length;j ++) {
                memo[i][j] = Integer.MAX_VALUE;
            }
        return dp(amount, coins, 0, memo);
    }

    private int dp(int amount, int[] coins, int k, int[][] memo) {
        if (amount == 0) {
            return 1;
        }
        if (k == coins.length) {
            return 0;
        }
        if (k == coins.length - 1) {
            return amount % coins[k] == 0 ? 1 : 0;
        }

        if (memo[amount][k] != Integer.MAX_VALUE) {
            return memo[amount][k];
        }

        int currCoin = coins[k];
        int numberMax = amount / currCoin;
        int way = 0;
        for (int i =0; i<=numberMax;i++) {
            var subproblem = dp(amount - currCoin * i, coins, k + 1, memo);
            way = way + subproblem;
        }
        memo[amount][k] = way;
        return memo[amount][k];
    }

}

File: python/medium/_518_coin_change_2/solution.py
from typing import List

class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        n = len(coins)
        dp = [[0 for _ in range(amount + 1)] for _ in range(n + 1)]
        # base case - have 1 way to change 0 amount
        for i in range(n+1):
            dp[i][0] = 1

        for c in range(1, n+1):
            for a in range(1, amount + 1):
                coin_value = coins[c - 1]

                # not using the i-th coin, only use the first (i-1)th coin
                dp[c][a] = dp[c - 1][a]
                # can use coin_value, unlimited times
                if a >= coin_value:
                    dp[c][a] += dp[c][a - coin_value]

        return dp[n][amount]